lib/Sisimai/RFC3464.pm

```diff
@@ -3,410 +3,289 @@ use v5.26;
+    state $boundaries = [
+        # When the new value added, the part of the value should be listed in $delimiters variable
+        # defined at Sisimai::RFC2045->makeFlat() method
+        "Content-Type: message/rfc822",
+        "Content-Type: text/rfc822-headers",
+        "Content-Type: message/partial",
+        "Content-Disposition: inline", # See lhost-amavis-*.eml, lhost-facebook-*.eml
+    ];
+    state $startingof = {"message" => ["Content-Type: message/delivery-status"]};
+    state $fieldtable = Sisimai::RFC1894->FIELDTABLE;
+
+    unless( grep { index($$mbody, $_) > 0 } @$boundaries ) {
+        # There is no "Content-Type: message/rfc822" line in the message body
+        # Insert "Content-Type: message/rfc822" before "Return-Path:" of the original message
+        my $p0 = index($$mbody, "\n\nReturn-Path:");
+        $$mbody = sprintf("%s%s%s", substr($$mbody, 0, $p0), $boundaries->[0], substr($$mbody, $p0 + 1,)) if $p0 > 0;
+    }
 
+    my $permessage = {};
     my $dscontents = [Sisimai::Lhost->DELIVERYSTATUS];
-    my $rfc822text = '';    # (String) message/rfc822 part text
-    my $maybealias = '';    # (String) Original-Recipient field
-    my $lowercased = '';    # (String) Lowercased each line of the loop
-    my $blanklines = 0;     # (Integer) The number of blank lines
+    my $alternates = Sisimai::Lhost->DELIVERYSTATUS;
+    my $emailparts = Sisimai::RFC5322->part($mbody, $boundaries);
     my $readcursor = 0;     # (Integer) Points the current cursor position
     my $recipients = 0;     # (Integer) The number of 'Final-Recipient' header
-    my $itisbounce = 0;     # (Integer) Flag for that an email is a bounce
-    my $connheader = {
-        'date'    => '',    # The value of Arrival-Date header
-        'rhost'   => '',    # The value of Reporting-MTA header
-        'lhost'   => '',    # The value of Received-From-MTA header
-    };
+    my $beforemesg = "";    # (String) String before $startingof->{"message"}
+    my $goestonext = 0;     # (Bool) Flag: do not append the line into $beforemesg
+    my $isboundary = [Sisimai::RFC2045->boundary($mhead->{"content-type"}, 0)]; $isboundary->[0] ||= "";
     my $v = undef;
-    my $p = '';
+    my $p = "";
+
+    while( index($emailparts->[0], '@') < 0 ) {
+        # There is no email address in the first element of emailparts
+        # There is a bounce message inside of message/rfc822 part at lhost-x5-*
+        my $p0 = -1;    # The index of the boundary string found first
+        my $p1 =  0;    # Offset position of the message body after the boundary string
+        my $ct = "";    # Boundary string found first such as "Content-Type: message/rfc822"
+
+        for my $e ( @$boundaries ) {
+            # Look for a boundary string from the message body
+            $p0 = index($$mbody, $e."\n"); next if $p0 < 0;
+            $p1 = $p0 + length($e) + 2;
+            $ct = $e; last;
+        }
+        last if $p0 < 0;
+
```

Comment:
- `my $cx = substr($$mbody, p1,)`
- And replace `substr($$mbody, p1,)` with `$cx`